### PR TITLE
Remove `onConnect` method from Bridge

### PR DESCRIPTION
**Pull request by robertknight**
Simplify the Bridge class by removing the `onConnect` API and replacing
current calls with alternative approaches.

The `onConnect` method was a way to register a callback that would be invoked
after a test RPC call over a newly-connected channel succeeded.

The purpose of this method was not that clear however and its existence
suggested that it is necessary to wait for the callback to be invoked before
making RPC requests on the new channel. This is not the case, at least not any more, as the underlying
transport buffers messages until the receiver is ready to consume them. Therefore
RPC requests can be made as soon as `createChannel` returns.

The current uses of `onConnect` were replaced as follows:

 1. The host frame used it to be notified when the sidebar application has
    loaded. This has been replaced with an explicit `ready` RPC call.

 2. The sidebar waited for this callback before querying metadata for a
    newly-connected guest. It now sends a metadata query to the guest
    immediately when it receives the guest port from the host frame. This saves
    an unnecessary round-trip between the host and guest. In future we can
    optimize metadata querying more by having the guest _push_ metadata to
    the sidebar rather than waiting for the sidebar to query it. See https://github.com/hypothesis/client/pull/4113.

Removing this API is a step towards combining the `Bridge` and `PortRPC` classes. See https://github.com/hypothesis/client/pull/4111.

This is part of https://github.com/hypothesis/client/issues/4094.

### CODE REVIEW

File: src/sidebar/services/test/frame-sync-test.js [hunk 1]
```diff
@@ -419,25 +445,19 @@ describe('FrameSyncService', () => {
   });
 
   context('when a new frame connects', () => {
```
Comment (esanzgar) [inline src/sidebar/services/test/frame-sync-test.js:447]:
```suggestion
  context('when a new guest frame connects', () => {
```

**Review by esanzgar — APPROVED** [review]
This PR allows to save a few milliseconds round-trip at the cost of stripping some generality. However, it is a step necessary towards merging `PortRPC` and `Bridge` classes.

I would like that `Guest` is responsible of sending the document metadata to the sidebar frame. That will simplify and speed up the process even further.

If we do the above, I believe the JSON-RPC protocol is strictly not needed anymore. However, we might want keep it for in case of future unforeseen needs.  

Other events, like `setHihlightsVisible` or `closeSidebar`, are transmitted to the sidebar frame, as an intermediate step, to then be re-broadcasted to the host or guests. A more direct communication `host-guest` makes this intermediate step unnecessary and the communication more intuitive and faster.

**Comment by robertknight** [comment]
> If we do the above, I believe the JSON-RPC protocol is strictly not needed anymore. However, we might want keep it for in case of future unforeseen needs.

Yes, I think we should keep the RPC protocol. We might not currently be using return values, but we'd still need to call methods with arguments, and it seems likely that the ability to return results or report errors will come in useful in future. A change we might want to make is to make the protocol actually JSON-RPC, like we use for communication with the LMS parent frame, rather than just being JSON-RPC-like. This might allow us to use a single implementation.

> Other events, like setHihlightsVisible or closeSidebar, are transmitted to the sidebar frame, as an intermediate step, to then be re-broadcasted to the host or guests. A more direct communication host-guest makes this intermediate step unnecessary and the communication more intuitive and faster

I agree.

File: src/sidebar/services/test/frame-sync-test.js [hunk 2]
```diff
@@ -60,27 +60,39 @@ describe('FrameSyncService', () => {
...
+        createChannel: sinon.stub().callsFake(() => {
+          const rpc = {
+            call: sinon.stub(),
+            destroy: sinon.stub(),
+          };
+
+          setupChannel?.(rpc);
+
+          return rpc;
+        }),
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/services/test/frame-sync-test.js:93]: Yes. It gets a lot simpler in https://github.com/hypothesis/client/pull/4111/ though.

**Pull request merged**